File: src/main/java/de/omilke/bankingfx/report/savings/model/Entry.java
package de.omilke.bankingfx.report.savings.model;

import javafx.beans.property.*;

import java.math.BigDecimal;
import java.time.LocalDate;

/**
 * @author Oliver Milke
 */
public class Entry {

    private final ObjectProperty<LocalDate> entryDate;
    private final ObjectProperty<BigDecimal> amount;
    private final BooleanProperty saving;
    private final StringProperty comment;

    private final String category;

    private final StringProperty groupLabel;

    public Entry(final String groupLabel, BigDecimal sum) {

        this.entryDate = new SimpleObjectProperty<>(null);
        this.saving = new SimpleBooleanProperty();
        this.comment = new SimpleStringProperty(null);

        this.category = null;

        this.groupLabel = new SimpleStringProperty(groupLabel);
        this.amount = new SimpleObjectProperty<>(sum);
    }

    public Entry() {

        this(null, null, false, null, null);
    }

    public Entry(final LocalDate entryDate, final BigDecimal amount, final boolean saving, final String category, final String comment) {

        this.entryDate = new SimpleObjectProperty<>(entryDate);
        this.amount = new SimpleObjectProperty<>(amount);
        this.saving = new SimpleBooleanProperty(saving);
        this.comment = new SimpleStringProperty(comment);

        this.category = category;

        this.groupLabel = new SimpleStringProperty(null);
    }

    public LocalDate getEntryDate() {

        return entryDate.get();
    }

    public ObjectProperty<LocalDate> entryDateProperty() {

        return entryDate;
    }

    public BigDecimal getAmount() {

        return amount.get();
    }

    public ObjectProperty<BigDecimal> amountProperty() {

        return amount;
    }

    public BooleanProperty savingProperty() {

        return saving;
    }

    public String getComment() {

        return comment.get();
    }

    public StringProperty commentProperty() {

        return comment;
    }

    public String getCategory() {

        return category;
    }

    public StringProperty groupLabelProperty() {

        return groupLabel;
    }

    public static Entry fromStorageModel(de.omilke.banking.account.entity.Entry current) {

        return new Entry(current.getEntryDate(), current.getAmount(), current.isSaving(), current.getCategory(), current.getComment());
    }
}

File: src/main/java/de/omilke/bankingfx/report/savings/model/Entry.kt
package de.omilke.bankingfx.report.savings.model

import javafx.beans.property.*
import java.math.BigDecimal
import java.time.LocalDate
import de.omilke.banking.account.entity.Entry as domainEntry

/**
 * @author Oliver Milke
 */
class Entry {

    val category: String?

    private val entryDate: ObjectProperty<LocalDate?>
    private val amount: ObjectProperty<BigDecimal?>
    private val saving: ObjectProperty<Boolean?>
    private val comment: StringProperty

    private val groupLabel: StringProperty

    constructor(entry: domainEntry)
            : this(entry.entryDate, entry.amount, entry.isSaving, entry.category, entry.comment)

    constructor(groupLabel: String?, sum: BigDecimal?)
            : this(null, sum, null, null, null, groupLabel)

    constructor(entryDate: LocalDate? = null, amount: BigDecimal? = null, saving: Boolean? = false, category: String? = null, comment: String? = null, groupLabel: String? = null) {

        this.category = category

        this.entryDate = SimpleObjectProperty(entryDate)
        this.amount = SimpleObjectProperty(amount)
        this.saving = SimpleObjectProperty(saving)
        this.comment = SimpleStringProperty(comment)
        this.groupLabel = SimpleStringProperty(groupLabel)
    }

    fun getEntryDate(): LocalDate? {
        return entryDate.get()
    }

    fun entryDateProperty(): ObjectProperty<LocalDate?> {
        return entryDate
    }

    fun getAmount(): BigDecimal? {
        return amount.get()
    }

    fun amountProperty(): ObjectProperty<BigDecimal?> {
        return amount
    }

    fun savingProperty(): ObjectProperty<Boolean?> {
        return saving
    }

    fun getComment(): String {
        return comment.get()
    }

    fun commentProperty(): StringProperty {
        return comment
    }

    fun groupLabelProperty(): StringProperty {
        return groupLabel
    }

}

